File: C++/1088.cpp
__________________________________________________________________________________________________
class Solution {
public:
    int f[10] = {0, 1, 0, 0, 0, 0, 9, 0, 8, 6};
    int N, len, ans = 0;
    int bit[5] = {0, 1, 6, 8, 9};
    int a[20];
    int check(int x, int m){
        int l = 0;
        while(l < m && a[l] == 0) ++l;
        if(l == m) return 0;
        for(int i = l; i <= l + (m - l) / 2; ++i){
            if(f[a[i]] != a[m - 1 - (i - l)]){
                return 1;
            }
        }
        return 0;
    }
    void go(long long x, int m){
        if(x > N) return ;
        if(m > len){
            ans += check(x, m);
            return ;
        }
        for(int i = 0; i < 5; ++i){
            a[m] = bit[i];
            go(x * 10 + bit[i], m + 1);
        }
    }
    int confusingNumberII(int nn) {
        N = nn;
        for(; nn; nn /= 10) ++len;
        go(0, 0);
        return ans;
    }
};

__________________________________________________________________________________________________

__________________________________________________________________________________________________

File: Java/1088.java
__________________________________________________________________________________________________
class Solution {
    public int confusingNumberII(int N) {
        int[] b = {0, 1, 6, 8, 9};
        int[] c = {0, 1, 9, 8, 6};
        
        int ans = 0;
        if(N == 1000000000)ans++;
        for(int d = 1;d <= 9;d++){
            int[] a = new int[d];
            do{
                if(a[0] == 0)continue;
                int or = 0;
                int ro = 0;
                for(int i = 0;i < d;i++){
                    or = or * 10 + b[a[i]];
                }
                for(int i =d-1;i >= 0;i--){
                    ro = ro * 10 + c[a[i]];
                }
                if(or <= N && or != ro){
                    ans++;
                }
            }while(inc(a, 5));
        }
        return ans;
    }
    
    public boolean inc(int[] a, int base) {
        int n = a.length;
        int i;
        for (i = n - 1; i >= 0 && a[i] == base - 1; i--)
            ;
        if (i == -1)
            return false;

        a[i]++;
        Arrays.fill(a, i + 1, n, 0);
        return true;
    }
}

__________________________________________________________________________________________________

__________________________________________________________________________________________________
